pkg/kubelet/nvidiagpu/nvidia_gpu_manager.go

```diff
@@ -0,0 +1,187 @@
+
+	// The interface which could get GPU mapping from all the containers.
+	// TODO: Should make this independent of Docker in the future.
+	dockerClient dockertools.DockerInterface
+}
+
+// Get all the paths of NVIDIA GPU card from /dev/
+// TODO: Without NVML support we only can check whether there has GPU devices, but
+// could not give a health check or get more information like GPU cores, memory, or
+// family name. Need to support NVML in the future. But we do not need NVML until
+// we want more features, features like schedule containers according to GPU family
+// name.
+func (ngm *NvidiaGPUManager) discovery() error {
+	var err error
+	if ngm.gpuPaths == nil {
+		err = filepath.Walk("/dev", func(path string, f os.FileInfo, err error) error {
+			reg := regexp.MustCompile(`^nvidia[0-9]*$`)
+			gpupath := reg.FindAllString(f.Name(), -1)
+			if gpupath != nil && gpupath[0] != "" {
+				ngm.gpuPaths = append(ngm.gpuPaths, "/dev/"+gpupath[0])
+			}
+
+			return nil
+		})
+	}
+
+	return err
+}
+
+func Valid(path string) bool {
+	reg := regexp.MustCompile(`^/dev/nvidia[0-9]*$`)
+	check := reg.FindAllString(path, -1)
+
+	return check != nil && check[0] != ""
+}
+
+// Initial the GPU devices, so far only need to discover the GPU paths.
+func (ngm *NvidiaGPUManager) Init(dc dockertools.DockerInterface) error {
+	if _, err := os.Stat(NvidiaDeviceCtl); err != nil {
+		return err
+	}
+
+	if _, err := os.Stat(NvidiaDeviceUVM); err != nil {
+		return err
+	}
+
+	ngm.gpuMutex.Lock()
+	defer ngm.gpuMutex.Unlock()
+
+	err := ngm.discovery()
+
+	ngm.dockerClient = dc
+
+	return err
+}
+
+func (ngm *NvidiaGPUManager) Shutdown() {
+	ngm.gpuMutex.Lock()
+	defer ngm.gpuMutex.Unlock()
+
+	ngm.gpuPaths = nil
+}
+
+// Get how many GPU cards we have.
+func (ngm *NvidiaGPUManager) Capacity() int {
+	ngm.gpuMutex.Lock()
+	defer ngm.gpuMutex.Unlock()
+
+	return len(ngm.gpuPaths)
+}
+
+// Check whether the GPU device could be assigned to a container.
+// Only check alive containers.
+func (ngm *NvidiaGPUManager) isAvailable(path string) bool {
+	containers, err := ngm.dockerClient.ListContainers(dockertypes.ContainerListOptions{All: true})
+
+	if err != nil {
+		return true
+	}
+
+	for i := range containers {
+		containerJSON, err := ngm.dockerClient.InspectContainer(containers[i].ID)
+		if err != nil {
+			continue
+		}
+
+		devices := containerJSON.HostConfig.Devices
+		if devices == nil {
+			continue
+		}
+
+		status := containerJSON.State.Status
+		if status == "exited" || status == "dead" {
+			continue
+		}
+
+		for _, device := range devices {
+			if Valid(device.PathOnHost) && path == device.PathOnHost {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// Return the GPU paths as needed, otherwise, return error.
+func (ngm *NvidiaGPUManager) AllocateGPUs(num int) (paths []string, err error) {
+	if num <= 0 {
+		return
+	}
+
+	ngm.gpuMutex.Lock()
+	defer ngm.gpuMutex.Unlock()
+
+	for _, path := range ngm.gpuPaths {
+		if ngm.isAvailable(path) {
```

Comment:
> GenerateRunContainerOptions/generateContainerConfig/startContainer run serially.

I think it's a bad idea of making those operations serially. It will slow down container spawning without concurrency.

Could we make it another way, e.g. only makes GPU devices allocation serially or add some locks to it?